### PR TITLE
Added prompt to Sandcastle navigation.

### PR DESCRIPTION
**Pull request by ayk115**
Resolves issue #2178 . Storing html and js in variables whenever any demo loads. If the user tries to navigate:
- If the codes are edited, a prompt shows up.
- If the codes are unedited, navigation successful.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks @ayk115.

Can you please make this pass JSHint.  It currently fails with:
```
[runJsHint] Apps/Sandcastle/CesiumSandcastle.js: line 492, col 6, Missing semicolon.
```
See [here](https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/53688394).

@emackey do you want to review this?

**Comment by ayk115** [comment]
Done.
How can I run this test locally?

**Comment by pjcozzi** [comment]
JSHint will run automatically in Eclipse.  You can also run it from the command line (along with the actual Cesium tests) using Ant.  See the [Contributor's Guide](https://github.com/AnalyticalGraphicsInc/cesium/wiki/Contributor's-Guide#buildtargets).

File: Apps/Sandcastle/CesiumSandcastle.js [hunk 1]
```diff
@@ -479,6 +481,16 @@ require({
...
     });
 
+    window.onbeforeunload = function (e) {
+        var htmlText = (htmlEditor.getValue()).replace(/\s/g, '');
+        var jsText = (jsEditor.getValue()).replace(/\s/g, '');
+        if (demoHtml === htmlText && demoJs === jsText) {
+            return ;
+        }
+
+        return 'Changes made by you are not saved.';
```
Comment (mramato) [inline Apps/Sandcastle/CesiumSandcastle.js]: This function would be cleaner if you negate the if statement and move the return into it.  I would also update the message to something like "You have unsaved changes. Are you sure you want to navigate away from this demo?".

```javascript
if (demoHtml !== htmlText || demoJs !== jsText) {
    return 'You have unsaved changes. Are you sure you want to navigate away from this demo?';
}
```

**Comment by mramato** [comment]
We should also make sure we set the `demoJs` and `demoHtml` to the content of the editors when the "Save html file" button is used, this way we don't prompt after a save.

**Comment by mramato** [comment]
@ayk115 In order to merge this we need you to fill out and email us a Contributor License Agreement. The details for this are in [Contributing.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md) You most likely want the [individual CLA](http://www.agi.com/licenses/individual-cla-agi-v1.0.txt)

**Comment by mramato** [comment]
@ayk115 Are you still planning on working on this?  I don't believe we've received your CLA yet and you haven't pushed any new commits in the last 2 weeks.  You can just push new commits directly to this branch to address our review comments.  Please let us know either way.  Thanks.

**Comment by ayk115** [comment]
@mramato I just signed the CLA and emailed it to cla@agi.com.  I have a slight doubt regarding the "Save html file" button. When the user clicks the button, a save file dialog appears. The user can choose to save the file or close the dialog. In the latter case, the file isn't actually saved and so, the `demoHtml` and `demoJs` variables shouldn't be set. The problem I now face is that how do I track whether the user has actually saved the file or closed the file dialog without saving? I couldn't figure out a way to access the save file dialog.
The easier solution is to set the variables when the user clicks the "Save html file"  button. In that case, we aren't actually checking the download action. Which solution should I implement?

**Comment by mramato** [comment]
@ayk115 Thanks, we have your CLA.  You make a good point about not being able to detect if save was canceled or not. It might be okay to just always assume they saved the file, since the alternative is to always prompt (which itself could be confusing) @emackey do you have any thoughts on this?

Assuming we just decide to always prompt, the only thing left (I think) is to perform the minor cleanup I mentioned above.

**Comment by emackey** [comment]
@ayk115 I think you're right, there's no way to tell if they saved it, so the safest thing is to always prompt.

**Comment by emackey** [comment]
In addition to @mramato's suggested cleanup, @ayk115 can you change the prompt text to this please?  I've tried to re-word it to be acceptable even if they've already saved their edits unknown to us.

```
Be sure to save a copy of any important edits before leaving this page.
Are you sure you want to navigate away from this demo?
```

**Comment by emackey** [comment]
Also we don't get a warning message if the user clicks on a different demo.  This won't trigger the `window.onbeforeunload` event, because we're not unloading Sandcastle, just loading a different demo.  I think you'll need to use `window.confirm` to get confirmation before loading the next demo.

**Comment by ayk115** [comment]
@mramato @emackey I did the changes you guys suggested. Please have a look.

**Comment by emackey** [comment]
Looks great, thanks!

**Comment by emackey** [comment]
Minor nitpick, there were some tabs instead of spaces in there, I fixed them before merging.  Thanks again!

**Pull request merged**